### ISSUE TITLE
`prisma migrate deploy`: `MariaDB doesn't yet support 'GET_LOCK in cluster (WSREP_ON=ON)'`

### ISSUE THREAD

**Issue by valerii15298** [prisma/prisma#19350]
### Bug description

When I am using `prisma migrate deploy` I get the following error:
```
failed: pnpm exec prisma migrate deploy
Error: This version of MariaDB doesn't yet support 'GET_LOCK in cluster (WSREP_ON=ON)'
   0: schema_core::state::ApplyMigrations
             at schema-engine/core/src/state.rs:197
```

### How to reproduce

It is production environment so cannot specify repro steps for which I do not have direct access...

### Expected behavior

Migrations should be applied successfully

### Prisma information

only using `prisma migrate deploy`

### Environment & setup

- OS: linux
- Database: MariaDB
- Node.js version: not relevant

### Prisma Version

```
prisma                  : 4.13.0
@prisma/client          : 4.13.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 1e7af066ee9cb95cf3a403c78d9aab3e6b04f37a (at ../node_modules/.pnpm/@prisma+engines@4.13.0/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 1e7af066ee9cb95cf3a403c78d9aab3e6b04f37a (at ../node_modules/.pnpm/@prisma+engines@4.13.0/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Format Wasm             : @prisma/prisma-fmt-wasm 4.13.0-50.1e7af066ee9cb95cf3a403c78d9aab3e6b04f37a
Default Engines Hash    : 1e7af066ee9cb95cf3a403c78d9aab3e6b04f37a
Studio                  : 0.484.0
```

**Comment by aqrln** [comment]
What version of MariaDB are you using?

**Comment by widders** [comment]
It's a 3 node MariaDB 10.6.12 Galera cluster but I don't think that's going to specifically be related, this is a limitation from the Galera clustering that locks are not usable. https://mariadb.com/kb/en/mariadb-galera-cluster-known-limitations/

Looks like your advisory locks are failing? Is there a way to bypass them?

**Comment by jkomyno** [comment]
@valerii15298 @widders hi, this is most likely due to [this usage](https://github.com/prisma/prisma-engines/blob/c694003bd22d0140d0f7911d1ddbdc9d7394f28b/schema-engine/connectors/sql-schema-connector/src/flavour/mysql.rs#L71-L88) of `SELECT GET_LOCK(...)` in our migration engine. Currently there's no configurable way of bypassing advisory lock. One could however skip them entirely on MariaDB like we already do on PlanetScale. Of course, PlanetScale provides branching workflows that lower the chance of race conditions occurring, but MariaDB does not - so we should think twice before avoiding locks for every MariaDB user. Moreover, it looks like Galera Cluster isn't MariaDB specific, but that [locks are unavailable in any Galera cluster version](https://fromdual.com/limitations-of-galera-cluster), even the MySQL one.

This poses the question: how can Prisma determine that it's connected to a Galera cluster?

**Comment by electrical** [comment]
```
show variables like '%wsrep_on%';
or
show variables WHERE Variable_name='wsrep_on';
```
If this gives the value 'ON' it means that Galera clustering is in use.

**Comment by Jolg42** [comment]
I could not reproduce using this docker image somehow 🤔 
```
docker run --name mariadb -p 3306:3306 -e ALLOW_EMPTY_PASSWORD=yes \
  bitnami/mariadb-galera:10.1
```
(Same with latest version)

`wsrep_on` is `ON` but `migrate` deploy works as expected.

**Comment by Jolg42** [comment]
Do you have a better idea on how to reproduce this?

**Comment by widders** [comment]
Were you only running one database node? I'm pretty sure this will only present when you run a cluster. You might just be able to bootstrap the first node and test with that. We run this using bitnami's helm chart but per https://github.com/bitnami/containers/tree/main/bitnami/mariadb-galera#setting-up-a-multi-master-cluster
Bootstrap a first node with:

```
docker run -d --name mariadb-galera-0 \
  -e MARIADB_GALERA_CLUSTER_NAME=my_galera \
  -e MARIADB_GALERA_MARIABACKUP_USER=my_mariabackup_user \
  -e MARIADB_GALERA_MARIABACKUP_PASSWORD=my_mariabackup_password \
  -e MARIADB_ROOT_PASSWORD=my_root_password \
  -e MARIADB_GALERA_CLUSTER_BOOTSTRAP=yes \
  -e MARIADB_USER=my_user \
  -e MARIADB_PASSWORD=my_password \
  -e MARIADB_DATABASE=my_database \
  -e MARIADB_REPLICATION_USER=my_replication_user \
  -e MARIADB_REPLICATION_PASSWORD=my_replication_password \
  bitnami/mariadb-galera:latest
```


Start additional nodes with:

```
docker run -d --name mariadb-galera-1 --link mariadb-galera-0:mariadb-galera \
  -e MARIADB_GALERA_CLUSTER_NAME=my_galera \
  -e MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://mariadb-galera:4567,0.0.0.0:4567 \
  -e MARIADB_GALERA_MARIABACKUP_USER=my_mariabackup_user \
  -e MARIADB_GALERA_MARIABACKUP_PASSWORD=my_mariabackup_password \
  -e MARIADB_ROOT_PASSWORD=my_root_password \
  -e MARIADB_REPLICATION_USER=my_replication_user \
  -e MARIADB_REPLICATION_PASSWORD=my_replication_password \
  bitnami/mariadb-galera:latest
```

It's a pretty much the same problem as #16020 but setting permissive is not an option as this allows operations that can cause the cluster to desync or splitbrain.

**Comment by Jolg42** [comment]
Thanks! Somehow I still can't reproduce using this 🤔 

But I could reproduce with Percona in https://github.com/prisma/prisma/issues/16020#issuecomment-1693451258

**Comment by Jolg42** [comment]
@widders I have a PR ready, I'll give you a dev version here to try out next week once it's merged :shipit:

**Comment by Jolg42** [comment]
In https://github.com/prisma/prisma-engines/pull/4171 I added a new environment variable to disable the advisory locking.

Example values that disable the locking 
```
PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK=1
PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK=anything
```

Values that DO NOT disable the locking
```
PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK=
PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK=0
```

Note: This is now fixed in `5.3.0-dev.13` if you want to try it out. Let us know if it works like expected.
It will be released in the official 5.3.0 version on Tuesday, September, 12th.
